### PR TITLE
moving dropout from model_fn to init_model_fn

### PR DESCRIPTION
**Pull request by znado**
- Cleans up a lot of dropout related pains, including the slowdowns from having it in `model_fn`.
- Adds support for setting dropout to None in speech pytorch models
- Copies conformer pytorch reference submission into a deepspeech one for use in `tests/reference_algorithm_tests.py`.

### CODE REVIEW

File: algorithmic_efficiency/workloads/fastmri/fastmri_pytorch/models.py [hunk 1]
```diff
@@ -14,21 +14,27 @@
...
+    <hhttps://arxiv.org/pdf/1505.04597.pdf>`_.
+    """
 
   def __init__(self,
                in_chans: int = 1,
                out_chans: int = 1,
                chans: int = 32,
                num_pool_layers: int = 4,
-               dropout: float = 0.0) -> None:
+               dropout: Optional[float] = 0.0) -> None:
```
Comment (runame) [inline algorithmic_efficiency/workloads/fastmri/fastmri_pytorch/models.py:29]: Nit: we could switch to `dropout_rate` for consistency.

File: algorithmic_efficiency/workloads/fastmri/fastmri_pytorch/workload.py [hunk 2]
```diff
@@ -120,16 +125,12 @@ def model_fn(
       model_state: spec.ModelAuxiliaryState,
       mode: spec.ForwardPassMode,
       rng: spec.RandomState,
-      dropout_rate: Optional[float],
-      aux_dropout_rate: Optional[float],
       update_batch_norm: bool) -> Tuple[spec.Tensor, spec.ModelAuxiliaryState]:
     del model_state
     del rng
-    del aux_dropout_rate
```
Comment (runame) [inline algorithmic_efficiency/workloads/fastmri/fastmri_pytorch/workload.py:128]: I think we can add `del dropout_rate`.

File: algorithmic_efficiency/workloads/imagenet_vit/imagenet_jax/workload.py [hunk 3]
```diff
@@ -48,20 +52,14 @@ def model_fn(
       model_state: spec.ModelAuxiliaryState,
       mode: spec.ForwardPassMode,
       rng: spec.RandomState,
-      dropout_rate: Optional[float],
-      aux_dropout_rate: Optional[float],
       update_batch_norm: bool) -> Tuple[spec.Tensor, spec.ModelAuxiliaryState]:
     del model_state
-    del aux_dropout_rate
```
Comment (runame) [inline algorithmic_efficiency/workloads/imagenet_vit/imagenet_jax/workload.py:55]: I think we can add `del dropout_rate` here.

File: algorithmic_efficiency/workloads/imagenet_vit/imagenet_pytorch/workload.py [hunk 4]
```diff
@@ -48,16 +55,12 @@ def model_fn(
       model_state: spec.ModelAuxiliaryState,
       mode: spec.ForwardPassMode,
       rng: spec.RandomState,
-      dropout_rate: Optional[float],
-      aux_dropout_rate: Optional[float],
       update_batch_norm: bool) -> Tuple[spec.Tensor, spec.ModelAuxiliaryState]:
     del model_state
     del rng
-    del aux_dropout_rate
```
Comment (runame) [inline algorithmic_efficiency/workloads/imagenet_vit/imagenet_pytorch/workload.py:56]: I think we can add `del dropout_rate` here.

File: reference_algorithms/development_algorithms/librispeech_deepspeech/librispeech_pytorch/submission.py [hunk 5]
```diff
@@ -0,0 +1,104 @@
+"""Training algorithm track submission functions for LibriSpeech."""
+from typing import Dict, Iterator, List, Tuple
+
+import numpy as np
+import torch
+
+from algorithmic_efficiency import spec
+
+device = torch.device("cuda:0" if torch.cuda.is_available() else "cpu")
+ctc_loss = torch.nn.CTCLoss(blank=0, reduction="none")
```
Comment (runame) [inline reference_algorithms/development_algorithms/librispeech_deepspeech/librispeech_pytorch/submission.py:10]: Is this used anywhere?

File: reference_algorithms/development_algorithms/librispeech_deepspeech/librispeech_pytorch/submission.py [hunk 6]
```diff
@@ -0,0 +1,104 @@
+"""Training algorithm track submission functions for LibriSpeech."""
+from typing import Dict, Iterator, List, Tuple
+
+import numpy as np
+import torch
+
+from algorithmic_efficiency import spec
+
+device = torch.device("cuda:0" if torch.cuda.is_available() else "cpu")
```
Comment (runame) [inline reference_algorithms/development_algorithms/librispeech_deepspeech/librispeech_pytorch/submission.py:9]: Should be removed.

**Review by runame — APPROVED** [review]
LGTM!
